src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/ConnectionPool/ChannelDbConnectionPool.cs

```diff
@@ -111,9 +306,260 @@ public void TransactionEnded(Transaction transaction, DbConnectionInternal trans
+
+                                // Cancellation happens through the ReadAsync call, which will complete the task.
+                                awaiter.UnsafeOnCompleted(() => mres.Set());
+                                mres.Wait(CancellationToken.None);
+                                connection = awaiter.GetResult();
+                            }
+                            finally
+                            {
+                                SyncOverAsyncSemaphore.Release();
+                            }
+                        }
+
+                        if (connection != null && CheckConnection(connection))
+                        {
+                            PrepareConnection(owningConnection, connection);
+                            return connection;
+                        }
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        cancellationToken.ThrowIfCancellationRequested();
+                        Debug.Assert(finalToken.IsCancellationRequested);
+
+                        throw ADP.PooledOpenTimeout();
+                    }
+                    catch (ChannelClosedException)
+                    {
+                        //TODO: exceptions from resource file
+                        throw new Exception("The connection pool has been shut down.");
+                    }
+
+                    // Third pass
+                    //
+                    // Try again to get an idle connection or open a new one.
+                    // If we're here, our waiting attempt on the idle connection channel was released with a null
+                    // (or bad connection), or we're in sync mode. Check again if a new idle connection has appeared since we last checked.
+                    connection ??= GetIdleConnection();
+
+                    // We might have closed a connection in the meantime and no longer be at max capacity
+                    // so try to open a new connection and if that fails, loop again.
+                    connection ??= await OpenNewInternalConnection(owningConnection, userOptions, timeout, async, cancellationToken).ConfigureAwait(false);
+
+                    if (connection != null)
+                    {
+                        PrepareConnection(owningConnection, connection);
+                        return connection;
+                    }
+                }
+            }
+            finally
+            {
+                //TODO: log error
+            }
+        }
+
+        /// <summary>
+        /// Tries to read a connection from the idle connection channel.
+        /// </summary>
+        /// <returns>Returns true if a valid idles connection is found, otherwise returns false.</returns>
+        /// TODO: profile the inlining to see if it's necessary (comment copied from npgsql implementation)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
This is a bit of a code smell - usually we want to let the jit/compiler do what it does best w/o mucking around with it (especially since it might not do anything).